File: api/lua/doc/tag.lua
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at https://mozilla.org/MPL/2.0/.
--
-- SPDX-License-Identifier: MPL-2.0

---Tag management
---@module TagModule
local tag_module = {}

---@alias Layout
---| "MasterStack" # One master window on the left with all other windows stacked to the right.
---| "Dwindle" # Windows split in half towards the bottom right corner.
---| "Spiral" # Windows split in half in a spiral.
---| "CornerTopLeft" # One main corner window in the top left with a column of windows on the right and a row on the bottom.
---| "CornerTopRight" # One main corner window in the top right with a column of windows on the left and a row on the bottom.
---| "CornerBottomLeft" # One main corner window in the bottom left with a column of windows on the right and a row on the top.
---| "CornerBottomRight" # One main corner window in the bottom right with a column of windows on the left and a row on the top.

---Add tags to the specified output.
---
---### Examples
---    local op = output.get_by_name("DP-1")
---    if op ~= nil then
---        tag.add(op, "1", "2", "3", "4", "5") -- Add tags with names 1-5
---    end
---You can also pass in a table.
---    local tags = {"Terminal", "Browser", "Code", "Potato", "Email"}
---    tag.add(op, tags) -- Add tags with those names
---@tparam Output output The output you want these tags to be added to.
---@tparam string ... The names of the new tags you want to add.
---@see Output.add_tags
function tag_module.add(output, ...) end

---Toggle a tag on the specified output. If `output` isn't specified, toggle it on the currently focused output instead.
---
---### Example
---    -- Assuming all tags are toggled off...
---    local op = output.get_by_name("DP-1")
---    tag.toggle("1", op)
---    tag.toggle("2", op)
---    -- will cause windows on both tags 1 and 2 to be displayed at the same time.
---@tparam string name The name of the tag.
---@tparam ?Output output The output.
---@see Tag.toggle
function tag_module.toggle(name, output) end

---Switch to a tag on the specified output, deactivating any other active tags on it.
---If `output` is not specified, this uses the currently focused output instead.
---Alternatively, provide a tag object instead of a name and output.
---
---This is used to replicate what a traditional workspace is on some other Wayland compositors.
---
---### Examples
---    -- Switches to and displays *only* windows on tag `3` on the focused output.
---    tag.switch_to("3")
---@tparam string name The name of the tag.
---@tparam ?Output output The output.
---@see Tag.switch_to
function tag_module.switch_to(name, output) end

---Set a layout for the tag on the specified output. If no output is provided, set it for the tag on the currently focused one.
---Alternatively, provide a tag object instead of a name and output.
---
---### Examples
---    -- Set tag `1` on `DP-1` to the `Dwindle` layout
---    tag.set_layout("1", "Dwindle", output.get_by_name("DP-1"))
---
---    -- Do the same as above. Note: if you have more than one tag named `1` then this picks the first one.
---    local t = tag.get_by_name("1")[1]
---    tag.set_layout(t, "Dwindle")
---@tparam string name The name of the tag.
---@tparam Layout layout The layout.
---@tparam ?Output output The output.
---@see Tag.set_layout
function tag_module.set_layout(name, layout, output) end

---Get all tags on the specified output.
---
---### Example
---    local op = output.get_focused()
---    if op ~= nil then
---        local tags = tag.get_on_output(op) -- All tags on the focused output
---    end
---@tparam Output output
---@treturn Tag[]
---@see Output.tags
function tag_module.get_on_output(output) end

---Get all tags with this name across all outputs.
---
---### Example
---    -- Given one monitor with the tags "OBS", "OBS", "VSCode", and "Spotify"...
---    local tags = tag.get_by_name("OBS")
---    -- ...will have 2 tags in `tags`, while...
---    local no_tags = tag.get_by_name("Firefox")
---    -- ...will have `no_tags` be empty.
---@tparam string name The name of the tag(s) you want.
---@treturn Tag[]
function tag_module.get_by_name(name) end

---Get all tags across all outputs.
---
---### Example
---    -- With two monitors with the same tags: "1", "2", "3", "4", and "5"...
---    local tags = tag.get_all()
---    -- ...`tags` should have 10 tags, with 5 pairs of those names across both outputs.
---@treturn Tag[]
function tag_module.get_all() end

---Get the specified tag's name.
---
---### Example
---    -- Assuming the tag `Terminal` exists...
---    print(tag.name(tag.get_by_name("Terminal")[1]))
---    -- ...should print `Terminal`.
---@tparam Tag t
---@treturn string|nil
---@see Tag.name
function tag_module.name(t) end

---Get whether or not the specified tag is active.
---@tparam Tag t
---@treturn boolean|nil
---@see Tag.active
function tag_module.active(t) end

---Get the output the specified tag is on.
---@tparam Tag t
---@treturn Output|nil
---@see OutputModule.get_for_tag
---@see Tag.output
function tag_module.output(t) end

-----------------------------------------------------

---Tag objects
---@classmod Tag
local tag = {}

---Get this tag's active status.
---@treturn boolean|nil active `true` if the tag is active, `false` if not, and `nil` if the tag doesn't exist.
---@see TagModule.active
function tag:active() end

---Get this tag's name.
---@treturn string|nil name The name of this tag, or nil if it doesn't exist.
---@see TagModule.name
function tag:name() end

---Get this tag's output.
---@treturn Output|nil output The output this tag is on, or nil if the tag doesn't exist.
---@see TagModule.output
function tag:output() end

---Switch to this tag.
---@see TagModule.switch_to
function tag:switch_to() end

---Toggle this tag.
---@see TagModule.toggle
function tag:toggle() end

---Set this tag's layout.
---@tparam Layout layout
---@see TagModule.set_layout
function tag:set_layout(layout) end

File: api/rust_grpc/src/tag.rs
//! Tag management.
//!
//! This module allows you to interact with Pinnacle's tag system.
//!
//! # The Tag System
//! Many Wayland compositors use workspaces for window management.
//! Each window is assigned to a workspace and will only show up if that workspace is being
//! viewed. This is a find way to manage windows, but it's not that powerful.
//!
//! Instead, Pinnacle works with a tag system similar to window managers like [dwm](https://dwm.suckless.org/)
//! and, the window manager Pinnacle takes inspiration from, [awesome](https://awesomewm.org/).
//!
//! In a tag system, there are no workspaces. Instead, each window can be tagged with zero or more
//! tags, and zero or more tags can be displayed on a monitor at once. This allows you to, for
//! example, bring in your browsers on the same screen as your IDE by toggling the "Browser" tag.
//!
//! Workspaces can be emulated by only displaying one tag at a time. Combining this feature with
//! the ability to tag windows with multiple tags allows you to have one window show up on multiple
//! different "workspaces". As you can see, this system is much more powerful than workspaces
//! alone.
//!
//! # Configuration
//! `tag` contains the [`Tag`] struct, which allows you to add new tags
//! and get handles to already defined ones.
//!
//! These [`TagHandle`]s allow you to manipulate individual tags and get their properties.

use futures::{channel::mpsc::UnboundedSender, executor::block_on, future::BoxFuture};
use num_enum::TryFromPrimitive;
use pinnacle_api_defs::pinnacle::{
    output::v0alpha1::output_service_client::OutputServiceClient,
    tag::{
        self,
        v0alpha1::{
            tag_service_client::TagServiceClient, AddRequest, RemoveRequest, SetActiveRequest,
            SetLayoutRequest, SwitchToRequest,
        },
    },
};
use tonic::transport::Channel;

use crate::output::{Output, OutputHandle};

/// A struct that allows you to add and remove tags and get [`TagHandle`]s.
#[derive(Clone, Debug)]
pub struct Tag {
    channel: Channel,
    fut_sender: UnboundedSender<BoxFuture<'static, ()>>,
}

impl Tag {
    pub(crate) fn new(
        channel: Channel,
        fut_sender: UnboundedSender<BoxFuture<'static, ()>>,
    ) -> Self {
        Self {
            channel,
            fut_sender,
        }
    }

    fn create_tag_client(&self) -> TagServiceClient<Channel> {
        TagServiceClient::new(self.channel.clone())
    }

    fn create_output_client(&self) -> OutputServiceClient<Channel> {
        OutputServiceClient::new(self.channel.clone())
    }

    /// Add tags to the specified output.
    ///
    /// This will add tags with the given names to `output` and return [`TagHandle`]s to all of
    /// them.
    ///
    /// # Examples
    ///
    /// ```
    /// // Add tags 1-5 to the focused output
    /// if let Some(op) = output.get_focused() {
    ///     let tags = tag.add(&op, ["1", "2", "3", "4", "5"]);
    /// }
    /// ```
    pub fn add(
        &self,
        output: &OutputHandle,
        tag_names: impl IntoIterator<Item = impl Into<String>>,
    ) -> impl Iterator<Item = TagHandle> {
        let mut client = self.create_tag_client();
        let output_client = self.create_output_client();

        let tag_names = tag_names.into_iter().map(Into::into).collect();

        let response = block_on(client.add(AddRequest {
            output_name: Some(output.name.clone()),
            tag_names,
        }))
        .unwrap()
        .into_inner();

        response.tag_ids.into_iter().map(move |id| TagHandle {
            client: client.clone(),
            output_client: output_client.clone(),
            id,
        })
    }

    /// Get handles to all tags across all outputs.
    ///
    /// # Examples
    ///
    /// ```
    /// let all_tags = tag.get_all();
    /// ```
    pub fn get_all(&self) -> impl Iterator<Item = TagHandle> {
        let mut client = self.create_tag_client();
        let output_client = self.create_output_client();

        let response = block_on(client.get(tag::v0alpha1::GetRequest {}))
            .unwrap()
            .into_inner();

        response.tag_ids.into_iter().map(move |id| TagHandle {
            client: client.clone(),
            output_client: output_client.clone(),
            id,
        })
    }

    /// Get a handle to the first tag with the given name on `output`.
    ///
    /// If `output` is `None`, the focused output will be used.
    ///
    /// # Examples
    ///
    /// ```
    /// // Get tag "1" on output "HDMI-1"
    /// if let Some(op) = output.get_by_name("HDMI-1") {
    ///     let tg = tag.get("1", &op);
    /// }
    ///
    /// // Get tag "Thing" on the focused output
    /// let tg = tag.get("Thing", None);
    /// ```
    pub fn get<'a>(
        &self,
        name: impl Into<String>,
        output: impl Into<Option<&'a OutputHandle>>,
    ) -> Option<TagHandle> {
        let name = name.into();
        let output: Option<&OutputHandle> = output.into();
        let output_module = Output::new(self.channel.clone(), self.fut_sender.clone());

        self.get_all().find(|tag| {
            let props = tag.props();

            let same_tag_name = props.name.as_ref() == Some(&name);
            let same_output = props.output.is_some_and(|op| {
                Some(op.name)
                    == output
                        .map(|o| o.name.clone())
                        .or_else(|| output_module.get_focused().map(|o| o.name))
            });

            same_tag_name && same_output
        })
    }

    /// Remove the given tags from their outputs.
    ///
    /// # Examples
    ///
    /// ```
    /// let tags = tag.add(output.get_by_name("DP-1")?, ["1", "2", "Buckle", "Shoe"]);
    ///
    /// tag.remove(tags); // "DP-1" no longer has any tags
    /// ```
    pub fn remove(&self, tags: impl IntoIterator<Item = TagHandle>) {
        let tag_ids = tags.into_iter().map(|handle| handle.id).collect::<Vec<_>>();

        let mut client = self.create_tag_client();

        block_on(client.remove(RemoveRequest { tag_ids })).unwrap();
    }
}

/// A handle to a tag.
///
/// This handle allows you to do things like switch to tags and get their properties.
#[derive(Debug, Clone)]
pub struct TagHandle {
    pub(crate) client: TagServiceClient<Channel>,
    pub(crate) output_client: OutputServiceClient<Channel>,
    pub(crate) id: u32,
}

/// Various static layouts.
#[repr(i32)]
#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, TryFromPrimitive)]
pub enum Layout {
    /// One master window on the left with all other windows stacked to the right
    MasterStack = 1,
    /// Windows split in half towards the bottom right corner
    Dwindle,
    /// Windows split in half in a spiral
    Spiral,
    /// One main corner window in the top left with a column of windows on the right and a row on the bottom
    CornerTopLeft,
    /// One main corner window in the top right with a column of windows on the left and a row on the bottom
    CornerTopRight,
    /// One main corner window in the bottom left with a column of windows on the right and a row on the top.
    CornerBottomLeft,
    /// One main corner window in the bottom right with a column of windows on the left and a row on the top.
    CornerBottomRight,
}

impl TagHandle {
    /// Activate this tag and deactivate all other ones on the same output.
    ///
    /// This essentially emulates what a traditional workspace is.
    ///
    /// # Examples
    ///
    /// ```
    /// // Assume the focused output has the following inactive tags and windows:
    /// // "1": Alacritty
    /// // "2": Firefox, Discord
    /// // "3": Steam
    /// tag.get("2")?.switch_to(); // Displays Firefox and Discord
    /// tag.get("3")?.switch_to(); // Displays Steam
    /// ```
    pub fn switch_to(&self) {
        let mut client = self.client.clone();
        block_on(client.switch_to(SwitchToRequest {
            tag_id: Some(self.id),
        }))
        .unwrap();
    }

    /// Set this tag to active or not.
    ///
    /// While active, windows with this tag will be displayed.
    ///
    /// While inactive, windows with this tag will not be displayed unless they have other active
    /// tags.
    ///
    /// # Examples
    ///
    /// ```
    /// // Assume the focused output has the following inactive tags and windows:
    /// // "1": Alacritty
    /// // "2": Firefox, Discord
    /// // "3": Steam
    /// tag.get("2")?.set_active(true);  // Displays Firefox and Discord
    /// tag.get("3")?.set_active(true);  // Displays Firefox, Discord, and Steam
    /// tag.get("2")?.set_active(false); // Displays Steam
    /// ```
    pub fn set_active(&self, set: bool) {
        let mut client = self.client.clone();
        block_on(client.set_active(SetActiveRequest {
            tag_id: Some(self.id),
            set_or_toggle: Some(tag::v0alpha1::set_active_request::SetOrToggle::Set(set)),
        }))
        .unwrap();
    }

    /// Toggle this tag between active and inactive.
    ///
    /// While active, windows with this tag will be displayed.
    ///
    /// While inactive, windows with this tag will not be displayed unless they have other active
    /// tags.
    ///
    /// # Examples
    ///
    /// ```
    /// // Assume the focused output has the following inactive tags and windows:
    /// // "1": Alacritty
    /// // "2": Firefox, Discord
    /// // "3": Steam
    /// tag.get("2")?.toggle(); // Displays Firefox and Discord
    /// tag.get("3")?.toggle(); // Displays Firefox, Discord, and Steam
    /// tag.get("3")?.toggle(); // Displays Firefox, Discord
    /// tag.get("2")?.toggle(); // Displays nothing
    /// ```
    pub fn toggle_active(&self) {
        let mut client = self.client.clone();
        block_on(client.set_active(SetActiveRequest {
            tag_id: Some(self.id),
            set_or_toggle: Some(tag::v0alpha1::set_active_request::SetOrToggle::Toggle(())),
        }))
        .unwrap();
    }

    /// Remove this tag from its output.
    ///
    /// # Examples
    ///
    /// ```
    /// let tags = tag
    ///     .add(output.get_by_name("DP-1")?, ["1", "2", "Buckle", "Shoe"])
    ///     .collect::<Vec<_>>;
    ///
    /// tags[1].remove();
    /// tags[3].remove();
    /// // "DP-1" now only has tags "1" and "Buckle"
    /// ```
    pub fn remove(mut self) {
        block_on(self.client.remove(RemoveRequest {
            tag_ids: vec![self.id],
        }))
        .unwrap();
    }

    /// Set this tag's layout.
    ///
    /// Layouting only applies to tiled windows (windows that are not floating, maximized, or
    /// fullscreen). If multiple tags are active on an output, the first active tag's layout will
    /// determine the layout strategy.
    ///
    /// See [`Layout`] for the different static layouts Pinnacle currently has to offer.
    ///
    /// # Examples
    ///
    /// ```
    /// use pinnacle_api::tag::Layout;
    ///
    /// // Set the layout of tag "1" on the focused output to "corner top left".
    /// tag.get("1", None)?.set_layout(Layout::CornerTopLeft);
    /// ```
    pub fn set_layout(&self, layout: Layout) {
        let mut client = self.client.clone();
        block_on(client.set_layout(SetLayoutRequest {
            tag_id: Some(self.id),
            layout: Some(layout as i32),
        }))
        .unwrap();
    }

    /// Get all properties of this tag.
    ///
    /// # Examples
    ///
    /// ```
    /// use pinnacle_api::tag::TagProperties;
    ///
    /// let TagProperties {
    ///     active,
    ///     name,
    ///     output,
    /// } = tag.get("1", None)?.props();
    /// ```
    pub fn props(&self) -> TagProperties {
        let mut client = self.client.clone();
        let output_client = self.output_client.clone();

        let response = block_on(client.get_properties(tag::v0alpha1::GetPropertiesRequest {
            tag_id: Some(self.id),
        }))
        .unwrap()
        .into_inner();

        TagProperties {
            active: response.active,
            name: response.name,
            output: response.output_name.map(|name| OutputHandle {
                client: output_client,
                tag_client: client,
                name,
            }),
        }
    }

    /// Get this tag's active status.
    ///
    /// Shorthand for `self.props().active`.
    pub fn active(&self) -> Option<bool> {
        self.props().active
    }

    /// Get this tag's name.
    ///
    /// Shorthand for `self.props().name`.
    pub fn name(&self) -> Option<String> {
        self.props().name
    }

    /// Get a handle to the output this tag is on.
    ///
    /// Shorthand for `self.props().output`.
    pub fn output(&self) -> Option<OutputHandle> {
        self.props().output
    }
}

/// Properties of a tag.
pub struct TagProperties {
    /// Whether the tag is active or not
    pub active: Option<bool>,
    /// The name of the tag
    pub name: Option<String>,
    /// The output the tag is on
    pub output: Option<OutputHandle>,
}
